src/v/security/krb5.cc

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "security/krb5.h"
+
+#include "security/logger.h"
+#include "vlog.h"
+
+namespace security::krb5 {
+std::optional<context> context::initialize_krb5_context() noexcept {
+    ::krb5_context krb5_ctx{nullptr};
+    int krb5_rv = ::krb5_init_context(&krb5_ctx);
+    if (krb5_rv != 0) {
+        vlog(seclog.error, "Failed to initialize KRB5 instance: {}", krb5_rv);
+        return std::nullopt;
+    }
+
+    return context{krb5_ctx};
+}
+
+std::optional<default_realm>
+default_realm::get_default_realm(context_view ctx_view) noexcept {
+    char* realm{nullptr};
+    int krb5_rv = ::krb5_get_default_realm(*ctx_view, &realm);
+    if (krb5_rv != 0) {
+        vlog(seclog.error, "Failed to obtain default realm: {}", krb5_rv);
```

Comment:
Ideally the context of the error would be returned.

I thought about something like this:
```c++
namespace security::krb5 {
    template<typename T>
    using result = outcome_basic_result<t, ::krb5_error_code>;

    result<default_realm>
    default_realm::get_default_realm(context_view ctx_view) noexcept {
        char* realm{nullptr};
        int krb5_rv = ::krb5_get_default_realm(*ctx_view, &realm);
        if (krb5_rv != 0) {
            vlog(seclog.error, "Failed to obtain default realm: {}", krb5_rv);
            return krb5_rv;
        }

        return default_realm(ctx_view, realm);
    }
}
```
It should then be possible to get the error message outside of this function with `krb5_get_error_message`, but  

>  * The behavior of krb5_get_error_message() is only defined the first time it
>  * is called after a failed call to a krb5 function using the same context, and
>  * only when the error code passed in is the same as that returned by the krb5
>  * function.

So I guess it would be nice to just go ahead and do that:
```c++
namespace impl {
// Requires the context to outlive this object
class error_message {
    struct deleter {
        context_view ctx;
        void operator()(const char* realm) const {
            ::krb5_free_error_message(*ctx, realm);
        }
    };

public:
    explicit error_message(context_view ctx, const char* msg)
      : _msg{msg, deleter{ctx}} {}

    explicit operator std::string_view() { return _msg.get(); }

private:
    std::unique_ptr<const char, deleter> _msg;
};
struct error_impl {
    ::krb5_error_code ec{0};
    ss::sstring msg;
};

} // namespace impl

template<typename R>
using result = result<R, impl::error_impl>;

result<default_realm>
default_realm::get_default_realm(context_view ctx_view) noexcept {
    char* realm{nullptr};
    int krb5_rv = ::krb5_get_default_realm(*ctx_view, &realm);
    if (krb5_rv != 0) {
        impl::error_message msg{
          ctx_view, ::krb5_get_error_message(*ctx_view, krb5_rv)};
        auto err = impl::error_impl{
          krb5_rv, ss::sstring{std::string_view{msg}}};
        return err;
    }

    return default_realm(ctx_view, realm);
}
```
But it all seems like a lot of effort to shift the error handling by one stack frame.

Reply:
I think it's a lot easier to ignore the `result<>` stuff, but it might be worth at least seeing if the error message contains anything useful.